### ISSUE TITLE
Bug: linting + lsp errors when `jsx` is set to `preserve`

### ISSUE THREAD

**Issue by marvinhagemeister** [denoland/deno#25677]
Setting the `jsx` compiler option to `preserve` leads to an error in the LSP and when running `deno lint`. The `preserve` option is a valid value though.

## Steps to reproduce:

1. Create a `deno.json` with these contents:
  ```json
  {
    "compilerOptions": {
      "jsx": "preserve",
      "jsxImportSource": "preact"
    },
    "imports": {
      "preact": "npm:preact@^10.24.0"
    }
  }
  ```
2. Open the folder containing the `deno.json` in vscode.

Output:

```sh
Error caching: Unsupported 'jsx' compiler option value 'preserve'. Supported: 'react-jsx', 'react-jsxdev', 'react', 'precompile'
  at file:///Users/marvinh/dev/test/deno-jsx-preservec/deno.json
```

<img width="483" alt="Screenshot 2024-09-17 at 10 51 22" src="https://github.com/user-attachments/assets/6e15e368-f442-41fa-92ec-8a1b945ca705">

Version: Deno 2.0.0-rc.2+aaf2bf4

**Comment by lucacasonato** [comment]
When we pass the config to TSC, we should turn `preserve` into `react-jsx`.

**Comment by marvinhagemeister** [comment]
Updated the title as this also affects `deno lint`

**Comment by lucacasonato** [comment]
Actually - no, we don't support preserve anywhere. In the context of Deno, that option makes no sense because V8 would not be able to execute preserved JSX in a file. We always have to transpile it.

**Comment by marvinhagemeister** [comment]
Yeah, but when you use Deno in a vite project, the code will be transpiled by vite before handing it off to Deno.

**Comment by lucacasonato** [comment]
But for Deno (which is what the deno.json is configuring), `"preserve"` never makes sense

**Comment by marvinhagemeister** [comment]
But we need to configure it there for the LSP.

**Comment by lucacasonato** [comment]
Not in the `deno.json`? The `"preserve"` option for purposes of type checking behaves the same in the LSP as `"react"`, no?

**Comment by birkskyum** [comment]
> Not in the `deno.json`? The `"preserve"` option for purposes of type checking behaves the same in the LSP as `"react"`, no?

I'd be quite surprised if this is true, in cases like solid-js.

**Comment by lucacasonato** [comment]
Yes

**Comment by dsherret** [comment]
This is currently intended behaviour and not a bug. Probably we should improve the error message to explain why.

**Comment by lucacasonato** [comment]
I propose we enhance the error message for `"preserve"`, changing it to:

`Unsupported "jsx" compiler option value "preserve". Deno executes TypeScript instead of transpiling it to disk, so preserving JSX does not make sense. If you are setting "preserve" for a frontend framework like Solid, use "react-jsx" instead.`

**Comment by marvinhagemeister** [comment]
Disagree, we should support the `preserve` option or set it to `react-jsx` behind the scenes. There shouldn't be any worning. Not every JSX transpilation goes through us when bundlers are used. Setting it to `react-jsx` makes no sense for a SolidJS user.

**Comment by dsherret** [comment]
But the compiler options in a deno.json are for Deno and not a bundler. I like the idea of enhancing the error message and not supporting a preserve mode that's not really preserve.

**Comment by nayeemrmn** [comment]
> Setting it to `react-jsx` makes no sense for a SolidJS user.

Hmm.. from playing around with SolidJS templates and userland projects it seems like they are a bit misguided in suggesting `{ "jsx": "preserve" }`. They should be using `{ "jsx": "react-jsx", "noEmit": true }` which works fine. They don't rely on tsc's transpiler at all, why bother telling tsc to stop at `.tsx -> .jsx` transpilation?

That's what `preserve` does after all. In spirit it seems meant for bundlers that can handle `file.jsx` but not `file.tsx`, and makes tsc do only that half of the work. But bundlers like the ones solid works with all support TS.

SolidJS's [starter templates](https://github.com/solidjs/templates/tree/924ee8359dc908c3ca3bd622c0c32dd2d666c78e) seem to use `noEmit: true`, rendering `jsx: preserve` meaningless.

**Comment by birkskyum** [comment]
@nayeemrmn , if there are better ways to handle typescript for solid in deno than how it's done normally, I'm sure the community is quite open to suggestions. It's quite cursed if typescript has hardcoded some generic jsx behavior as "react", but that's not really deno's responsibility to clean up, and having the same ts settings across react/solid would in principle be  convenient, albeit the super confusing naming in tsconfig.

**Comment by nayeemrmn** [comment]
To be clear I'm not 100% sure that `preserve` is the same as `react-jsx` for typechecking, but it's got to align with at least one of the other options right?

**Comment by devdumpling** [comment]
Hi! Coming across this rabbit hole now in userland (have lots of experience with React tooling but new to Solid/deno) and wanted to add my 2c.

While adding the `react-jsx` option works fine (I think?) to fix the deno LSP, it's a confusing papercut that seems antithetical to Deno's (I think?) philosophy of lowering the required knowledge of JS ecosystem arcana to get going. I'm not sure what the right answer here is necessarily, either, but neither an error about `preserve` behavior nor requiring `react-jsx` in projects that aren't technically react feels great. Equally I get the concern around shadow supporting `preserve` in `deno.json` when it doesn't really make sense. 

I think what _feel_ better would be if Deno removed the need to configure this altogether behind the scenes. Like what if I could throw up my tsconfig with `preserve` and `solid-js` (to satisfy Vite, etc) but the Deno LSP natively understood that this is effectively still just JSX, just backed by a different import (Solid in my case). 

Maybe that's too magical, idk. Just a thought. 

I think barring that I personally would prefer the warning message concisely explaining "Oh hey btw preserve doesn't actually make sense in deno--you probably don't need this. See [link to explanation doc] for more information." At least then people could avoid this rabbit hole in the future, knowing that it's a design decision. 

The more challenging part to all of this is getting other vendors to include/understand the problem in their docs. I only arrived here in the first place because a Solid/Astro getting started guide had me put `preserve`. If you're interested, I give my breadcrumbs below.

----

My experience thus far has been as follows:

I'm building in Astro and added Solid for some interactive components. In following the [official Astro Solid integration guide](https://docs.astro.build/en/guides/integrations-guide/solid-js/#getting-started), I went the manual route as running `deno task astro add solid` fails (it tries to use `npm` instead of `deno`, which is just an unrelated Astro CLI papercut). 

In their docs they suggest adding the following to your `tsconfig.json`

```json
  "compilerOptions": {
    "jsx": "preserve",
    "jsxImportSource": "solid-js"
  }

```

Which mirrors the [Solid configuring typescript docs](https://docs.solidjs.com/configuration/typescript#configuring-typescript) where they state

> "jsx": "preserve" in the tsconfig.json retains the original JSX form. This is because Solid's JSX transformation is incompatible with TypeScript's JSX transformation.

This fails with an error (I'm using Zed but it's the same in VS Code) like this:

<img width="371" alt="Image" src="https://github.com/user-attachments/assets/798c16b8-7aa4-4e97-b182-eb6dd1ef3dc4" />

I came across the [Deno blog post/example for Solid](https://deno.com/blog/build-solidjs-with-deno) while trying to research this (after sifting through a few github issues unsuccessfully) which suggested the fix mentioned in this issue.

At first I thought, great! That works... until I noticed it was `react-jsx` and that struck me as off. Changing to `preserve` gave me this unhelpful error:

<img width="464" alt="Image" src="https://github.com/user-attachments/assets/2ea40f11-b5a4-4eab-bb7c-435a096e2b93" />

which ultimately led me to this issue. 

Hope that helps add some color! I'm still not exactly sure how this will all run, but so far using `react-jsx` is working. Will update if I run into other issues.

**Comment by wesleycoder** [comment]
Same here, I've set up a repo with solid-start via `deno init --npm solid` and for my editor (vscode) to be able to infer intrinsic elements correctly we need to use `"jsxImportSource": "solid-js"` which does trigger the LSP warning via the editor extension every time we change `deno.json` which turns out to be a bit spammy.

**Comment by nayeemrmn** [comment]
Fixed by #29843.
